datadoglog/formatters.py

```diff
@@ -1,47 +1,62 @@
-import os
 import json
 import logging
-import datetime
+import os
 import traceback
+from datetime import datetime
+from datetime import timezone
 
 
-class FormatFactory:
+class DatadogFormatFactory:
+    """
+    Formats our log record into a form that Datadog understands.
 
-    app_key = None
-    service = None
-    source = None
-    env = None
+    More information here:
+    https://docs.datadoghq.com/logs/log_configuration/attributes_naming_convention/
+    """
+
+    __slots__ = (
+        "app_key",
+        "env",
+        "host",
+        "service",
+        "source",
+    )
 
     def __init__(self, app_key: str, source: str, service: str, env: str) -> None:
         self.app_key = app_key
-        self.service = service
-        self.source = source
         self.env = env
         self.host = os.uname().nodename
+        self.service = service
+        self.source = source
 
     def my_format(self, record: logging.LogRecord) -> str:
+        timestamp = datetime.fromtimestamp(record.created, timezone.utc)
+
         data = {
-            "timestamp": datetime.datetime.fromtimestamp(
-                record.created, datetime.timezone.utc).strftime(
-                    "%Y-%m-%dT%H:%M:%S.%fZ"),
-            "host": self.host,
-            "ddsource": self.source,
-            "syslog.env": self.env,
-            "service": self.service,
-            "message": record.getMessage(),
-            "level": record.levelname,
+            # when
+            "syslog.timestamp": timestamp.strftime("%Y-%m-%dT%H:%M:%S.%fZ"),
+            # standard facets/attributes
+            "dd.env": self.env,
+            "dd.host": self.host,
+            "dd.service": self.service,
+            "dd.source": self.source,
+            # the log info and message
             "logger.name": record.name,
-            "pathname": record.pathname,
+            "message": record.getMessage(),
+            "status": record.levelname,
+            # where the log record came from
             "lineno": record.lineno,
+            "pathname": record.pathname,
         }
-        # work with exception here
-        ex_info = record.exc_info
-        if ex_info is not None:
-            (extype, value, tb) = ex_info
-            data["error.message"] = str(value)
-            data["error.kind"] = str(extype)
-            data["error.stack"] = "".join(traceback.format_tb(tb))
-        return "{} {}".format(self.app_key, json.dumps(data))
+        # work with exception info here
+        exc_info = record.exc_info
+        if exc_info:
+            (exc_type, exc_value, exc_tb) = exc_info
+            data["error.kind"] = str(exc_type)
+            data["error.message"] = str(exc_value)
+            data["error.stack"] = "\n".join(traceback.format_tb(exc_tb))
```

Comment:
I saw this linked from elsewhere and was just a bit curious 👃...

I think the prior version of this line was correct: to be doing `"".join(...)` rather than `"\n".join(...)`.

Perhaps double-check in the specific Python version you're expecting to use, but 3.12 docs say the following about the output of `format_list` (which is used by `format_tb`):

> [...] return a list of strings ready for printing. [...] **Each string ends in a newline;** the strings may contain internal newlines as well, for those items whose source text line is not None.

-- https://docs.python.org/3.12/library/traceback.html#traceback.format_list

---

This jumped out at me because I had also been logging tracebacks recently. In my case I was using `format_stack` cause I didn't have an exception, but same difference I think: `format_stack` also relies on `format_list`.

(Speculative: if we wind up hitting some DD length limits and the joined TB strings gets truncated, could also probably leave it as a list of strings.)

Reply:
Thanks for catching that, checked the docs for all versions 3.7+ and it's the same doc for the `format_list` method - I've reverted the change here: https://github.com/gadventures/datadoglog/pull/1/commits/fb38d28359821f50a9ad929e23e759071e7a0d03